packages/arb-token-bridge-ui/src/token-bridge-sdk/EnhancedProvider.ts

```diff
@@ -0,0 +1,181 @@
+import {
+  Networkish,
+  StaticJsonRpcProvider,
+  TransactionReceipt
+} from '@ethersproject/providers'
+import { BigNumber } from 'ethers'
+import { ChainId } from '../types/ChainId'
+import { ConnectionInfo } from 'ethers/lib/utils.js'
+
+type CachedReceipts = Record<string, TransactionReceipt>
+
+interface Storage {
+  getItem(key: string): string | null
+  setItem(key: string, value: string): void
+}
+
+class WebStorage implements Storage {
+  getItem(key: string): string | null {
+    return localStorage.getItem(key)
+  }
+
+  setItem(key: string, value: string): void {
+    localStorage.setItem(key, value)
+  }
+}
+
+const localStorageKey = `arbitrum:bridge:tx-receipts-cache`
+const enableCaching = true
+
+const getCacheKey = (chainId: number | string, txHash: string) =>
+  `${chainId}:${txHash}`.toLowerCase()
+
+/**
+ * Converts a TransactionReceipt to a string by encoding BigNumber properties.
+ * @param txReceipt - The receipt to encode.
+ * @returns The encoded receipt as a string.
+ */
+const txReceiptToString = (txReceipt: TransactionReceipt): string => {
+  const encodedReceipt: Record<string, any> = { ...txReceipt }
+
+  Object.keys(encodedReceipt).forEach(field => {
+    if (encodedReceipt[field]?._isBigNumber) {
+      encodedReceipt[field] = {
+        _type: 'BigNumber',
+        _data: encodedReceipt[field].toString()
+      }
+    }
+  })
+
+  return JSON.stringify(encodedReceipt)
+}
+
+/**
+ * Converts a stringified receipt back to a TransactionReceipt by decoding BigNumber properties.
+ * @param stringified - The stringified receipt to decode.
+ * @returns The decoded TransactionReceipt.
+ */
+const txReceiptFromString = (stringified: string): TransactionReceipt => {
+  const receipt = JSON.parse(stringified)
+  const decodedReceipt = { ...receipt }
+
+  Object.keys(decodedReceipt).forEach(field => {
+    if (decodedReceipt[field]?._type === 'BigNumber') {
+      decodedReceipt[field] = BigNumber.from(decodedReceipt[field]._data)
+    }
+  })
+
+  return decodedReceipt as TransactionReceipt
+}
+
+/**
+ * Checks if a transaction receipt can be cached based on its status and confirmations.
+ * @param chainId - The ID of the chain.
+ * @param txReceipt - The transaction receipt to check.
+ * @returns True if the receipt can be cached, false otherwise.
+ */
+const shouldCacheTxReceipt = (
+  chainId: number,
+  txReceipt: TransactionReceipt
+): boolean => {
+  if (!enableCaching) return false
+
+  // Don't cache failed transactions
+  if (txReceipt.status === 0) {
+    return false
+  }
```

Comment:
don't think we need this, we should cache also failed transactions, that's their terminal state